build_runner/lib/src/generate/options.dart

```diff
@@ -43,12 +49,24 @@ class BuildOptions {
       this.writer,
       this.skipBuildScriptCheck,
       this.enableLowResourcesMode}) {
-    /// Set up logging
+    // Set up logging
     logLevel ??= Level.INFO;
+
+    // Invalid to have Level.OFF but want severe logs to fail the build.
+    if (logLevel == Level.OFF && failOnSevere == true) {
+      logLevel = Level.SEVERE;
+    }
+
     Logger.root.level = logLevel;
 
     overrideAnsiOutput(assumeTty == true || ansiOutputEnabled, () {
-      logListener = Logger.root.onRecord.listen(onLog ?? stdIOLogListener);
+      onLog ??= stdIOLogListener;
+      logListener = Logger.root.onRecord.listen((r) {
+        if (r.level == Level.SEVERE) {
+          severeLogHandled = true;
```

Comment:
do we ever reset this (for watch mode)?

Reply:
Yup, right here:

https://github.com/dart-lang/build/pull/781/files#diff-7a296c2aaedb49f2b91b7fa9c81b536cR89

Reply:
And again in `watch_impl.dart`.